### PR TITLE
Stop Docker runtime in case of running runtime error

### LINKED ISSUE #5419: Stop Docker runtime in case of appearance of internal error like OOM
When some Runtime finds out that an internal error happened in the runtime and it can not continue its work it should stop the runtime and notify SPI to clear up WS status.

### PR DESCRIPTION
**Pull request by garagatyi**
This change is relevant to the spi branch only.
### What does this PR do?
Stop Docker runtime in case of running runtime error.
Move workspace statuses management to WorkspaceRuntimes.

### What issues does this PR fix or reference?
Fixes #5419

### CODE REVIEW

**Review by akorneta — APPROVED** [review]

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java [hunk 1]
```diff
@@ -249,48 +264,45 @@ public Runtime get(String workspaceId) throws NotFoundException, ServerException
...
                                                                              InfrastructureException,
                                                                              ConflictException {
-
+        RuntimeState state = runtimes.get(workspaceId);
+        if (state == null) {
+            throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+        }
+        if (!state.status.equals(RUNNING)) {
+            throw new ConflictException(
+                    format("Could not start workspace '%s' because its state is '%s'", workspaceId, state.status));
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: typo _start_ -> _stop_

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java [hunk 2]
```diff
@@ -249,48 +264,45 @@ public Runtime get(String workspaceId) throws NotFoundException, ServerException
...
+        }
+        runtimes.replace(workspaceId, new RuntimeState(state.runtime, STOPPING));
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: I'm not sure i understand the whole flow, but it looks like this is a place which should be passed by the one thread only. I think that this version of replace allows you to achieve such behaviour:
```java
if (!runtimes.replace(workspaceId, state, new RuntimeState(state.runtime, STOPPING))) {
    // some other thread already put the new value, handle it somehow
}
```
WDYT?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java [hunk 3]
```diff
@@ -301,11 +313,21 @@ public boolean hasRuntime(String workspaceId) {
...
     private void recover() {
         for (RuntimeInfrastructure infra : infraByRecipe.values()) {
             try {
                 for (RuntimeIdentity id : infra.getIdentities()) {
-                    runtimes.put(id.getWorkspaceId(), validate(infra.getRuntime(id)));
+                    // TODO how to identify correct state of runtime
+                    if (runtimes.putIfAbsent(id.getWorkspaceId(),
+                                             new RuntimeState(validate(infra.getRuntime(id)), RUNNING)) != null) {
+                        // should not happen, violation of SPI contract
+                        LOG.error("More than 1 runtime of workspace with id '' found. Skipping all duplicates.");
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: Missing _id_ parameter. I think more information should be logged, for example _RuntineIdentity_ of the duplicate

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java [hunk 4]
```diff
@@ -335,6 +361,34 @@ private static EnvironmentImpl copyEnv(Workspace workspace, String envName) {
...
+    private class CleanupRuntimeOnAbnormalRuntimeStop implements EventSubscriber<RuntimeStatusEvent> {
+        @Override
+        public void onEvent(RuntimeStatusEvent event) {
+            if (event.isFailed()) {
+                RuntimeState state = runtimes.remove(event.getIdentity().getWorkspaceId());
+                if (state != null) {
+                    eventService.publish(DtoFactory.newDto(WorkspaceStatusEvent.class)
+                                                   .withWorkspaceId(state.runtime.getContext().getIdentity()
+                                                                         .getWorkspaceId())
+                                                   .withPrevStatus(WorkspaceStatus.STOPPING)
```
Comment (voievodin) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: Are you sure it's _STOPPING_ ?
It looks like if the stop is abnormal then the flow should be _RUNNING -> STOPPED_

File: infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/monit/AbnormalMachineStopHandler.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.workspace.infrastructure.docker.monit;
+
+/**
+ * @author Alexander Garagatyi
+ */
+public interface AbnormalMachineStopHandler {
```
[the remark this answers is not in the document]
Comment (garagatyi) [reply infrastructures/docker/src/main/java/org/eclipse/che/workspace/infrastructure/docker/monit/AbnormalMachineStopHandler.java]: What if the handler doesn't use exception at all.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 1 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: tnx

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 2 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: There is no guarantee that objects will be equal. Some of the fields might change in the process of start, so it really depends on equal method implementation

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 3 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: ok

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 4 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: tnx

**Comment by garagatyi** [comment]
@evoevodin I updated PR in accordance with your comments

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java (hunk 2 above)
Comment (garagatyi) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java]: Yep, you are right. I updated the PR

**Review by voievodin — APPROVED** [review]

**Pull request merged**